zenoh/src/api/handlers/callback.rs

```diff
@@ -22,32 +25,88 @@ pub fn locked<T>(fnmut: impl FnMut(T)) -> impl Fn(T) {
     move |x| zlock!(lock)(x)
 }
 
-/// An immutable callback function.
-pub type Callback<'a, T> = Dyn<dyn Fn(T) + Send + Sync + 'a>;
+enum CallbackInner<T> {
+    Dyn(Arc<dyn Fn(&T) + Send + Sync>),
+    Flume(flume::Sender<T>),
+    Ring(RingChannelSender<T>),
+}
 
-impl<'a, T, F> IntoHandler<'a, T> for F
-where
-    F: Fn(T) + Send + Sync + 'a,
-{
+pub struct Callback<T>(CallbackInner<T>);
+
+impl<T> Clone for Callback<T> {
+    fn clone(&self) -> Self {
+        Self(match &self.0 {
+            CallbackInner::Dyn(cb) => CallbackInner::Dyn(cb.clone()),
+            CallbackInner::Flume(tx) => CallbackInner::Flume(tx.clone()),
+            CallbackInner::Ring(tx) => CallbackInner::Ring(tx.clone()),
+        })
+    }
+}
+
+impl<T> Callback<T> {
+    pub fn new(cb: Arc<dyn Fn(&T) + Send + Sync>) -> Self {
+        Self(CallbackInner::Dyn(cb))
+    }
+
+    pub(crate) fn new_flume(sender: flume::Sender<T>) -> Self {
+        Self(CallbackInner::Flume(sender))
+    }
+
+    pub(crate) fn new_ring(sender: RingChannelSender<T>) -> Self {
+        Self(CallbackInner::Ring(sender))
+    }
+
+    #[inline]
+    pub fn call(&self, arg: &T)
+    where
+        T: Clone,
+    {
+        match &self.0 {
+            CallbackInner::Dyn(cb) => cb(arg),
+            CallbackInner::Flume(tx) => {
```

Comment:
Are you sure that this kind of optimization is really justified given that flume/ring channels are just one of many that users can use in rust, and that many people are still going to just use function callbacks ? 
Also to make use of such an optimization we would need to complicate zenoh-c interface since right now it only accepts a "closure" with pointer to c-function (which is transformed into Arc<dyn>).

Reply:
Flume is not "just one of many that users can use in rust", it's the default one, and one of the few with a `IntoHandler` implementation (the other being `std:sync::mpsc`.
And the goal of this change is also to optimize callback use, see https://godbolt.org/z/G6vcT4b93

> Also to make use of such an optimization we would need to complicate zenoh-c interface since right now it only accepts a "closure" with pointer to c-function (which is transformed into Arc).

I don't see the how this PR changes anything here

Reply:
> I don't see the how this PR changes anything here

It adds an extra overhead due to enum match (not sure whether it is significant though). Which means that you probably improve performance of flume channel in rust, but maybe also make it worse for the whole zenoh-c/cpp + produce slightly more complex code. What I mean is that I would really like to see the performance gain for channels / loss for callbacks if there is any.

Reply:
> It adds an extra overhead due to enum match (not sure whether it is significant though).

Branch prediction enters the room.

> Which means that you probably improve performance of flume channel in rust, but maybe also make it worse for the whole zenoh-c/cpp + produce slightly more complex code.


Still better than the previous "by value" callback, see godbolt. Why is the produced code more complex?

> This also means that if in several month we decide to have another "default" channel we would need to revisit this code.

And then, we will change the internal enum, we don't care, it's internal, and it's only a dozen lines to change.

What I mean is that I would really like to see the performance gain for channels / loss for callbacks if there is any.

Trust branch prediction, especially if you are using *only* callbacks.

Reply:
> till better than the previous "by value" callback, see godbolt. Why is the produced code more complex?
> 
> >

I do not argue that passing by ref is faster than by value, this can still be achieved through only Arc<dyn>, without any enum matching. 
The produced code is more complex  since it adds more lines => higher maintenance cost (which is always better to be avoided if there is no any significant benefit)

Reply:
> But more than a performance optimization, the long-term vision is to be able to get flume sender directly in order to use its async method.

This is the goal behind making the type opaque. Now, using the enum is not a high maintenance code, all the code is contained in one module and is only a few lines.